### DISCUSSION TITLE
Integrate with Matrix

### DISCUSSION THREAD

**Discussion by IanLawrence** [outline/outline#1486] [General]
Really like Outline and what we see so far. We run a Matrix server and would really like to integate with Outline.
Please, please and please add onto the roadmap!

**Comment by matbrgz** [comment]
Which type of integration? Sending messages?

**Reply by IanLawrence** [reply]
Our dev team are all on matrix and we want to push threads to outline to start building a knowledge base

**Reply by tommoor** [reply]
It doesn't seem like this will be high on the roadmap because it's relatively unrequested – but this is why we have an open API for creating and updating documents, I'm sure a competent dev could put together an integration to push threads from Matrix in relatively little time if it would really speed up your workflow 😄

**Reply by amescher** [reply]
I also think that this is a cool feature. I'd like to fund this development and some more of outline. Is there a way I could do that?

**Reply by almereyda** [reply]
I see multiple options here how Outline can benefit from the Matrix ecosystem:

1. As a notification channel for changes in documents. (Outline > Matrix)  
  A bot that talks with the API, and is eventually triggered by Webhooks, could report those back into a channel that's linked to the workspace, as it is already done with email notifications.
2. As an aggregator for messages that are picked up by a bot, and then delivered to specific documents. (Matrix > Outline)
3. As a communication side channel that can be attached to documents ("comments", annotations, marginalia, discussions).

To further investigate such an integration, we can also ask ourselves:

- Which other integrations with text production and publication systems are known from Slack, Mattermost, Rocket.Chat et al.?

Personally, I would like to know @amescher, how is it that you sketch this feature for yourself, in case you were participating in a process that implements a bot for bidirectional integration of these systems?

**Comment by infogulch** [comment]
I think Matrix would be a good alternative to Slack / Microsoft / Google as an authentication provider and messaging platform. One benefit is that it enables an outline installation to be fully self hosted.

**Reply by IanLawrence** [reply]
We self host matrix and jitsi

**Reply by infogulch** [reply]
After considering further, perhaps a simple LDAP integration would be good enough, since you can set up Matrix to authenticate with LDAP as well. The irony of using "simple" and "LDAP" in the same sentence does not escape me 😁, though if the end goal stops at "authentication and identity management" then it's not too bad (it's more trouble when you start wanting to go crazy with setting dozens of user properties and syncing them).

And it seems there is already some feedback on LDAP integration. E.g. #809 https://github.com/outline/outline/projects/5

**Reply by amescher** [reply]
Wow. This is great.

**Reply by almereyda** [reply]
It seems the Matrix ecosystem is converging on LDAP, SAML and OIDC for authentication against the service, and will not target to be used as an independent IDP by itself. Therefore it makes much more sense to integrate Matrix and Outline with the same (federated) IDP in return.

**Reply by IanLawrence** [reply]
I would prefer OIDC given that it is JSON.
Love everything @almereyda mentioned especially the 'comments on matrix' idea. We are snowed under with work atm so unlikely to get round to it  until it becomes too painful not having the integration

**Comment by almereyda** [comment]
Which features from the Slack integration could we mimic for a first proof of concept?

This can also bring the basis for a decoupling of the notification/integration system around Slack, similarily to how authentication was decoupled from it.
